gpytorch/mlls/_approximate_mll.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+
+import torch
+from .marginal_log_likelihood import MarginalLogLikelihood
+from abc import ABC, abstractmethod
+from .. import settings
+
+
+class _ApproximateMarginalLogLikelihood(MarginalLogLikelihood, ABC):
+    r"""
+    An approximate marginal log likelihood (typically a bound) for approximate GP models.
+    We expect that :attr:`model` is a :obj:`gpytorch.models.ApproximateGP`.
+
+    Args:
+        :attr:`likelihood` (:obj:`gpytorch.likelihoods.Likelihood`):
+            The likelihood for the model
+        :attr:`model` (:obj:`gpytorch.models.ApproximateGP`):
+            The approximate GP model
+        :attr:`num_data` (int):
+            The total number of training data points (necessary for SGD)
+        :attr:`beta` (float - default 1.):
+            A multiplicative factor for the KL divergence term.
+            Setting it to 1 (default) recovers true variational inference
+            (as derived in `Scalable Variational Gaussian Process Classification`_).
+            Setting it to anything less than 1 reduces the regularization effect of the model
+            (similarly to what was proposed in `the beta-VAE paper`_).
+        :attr:`combine_terms` (bool):
+            Whether or not to sum the expected NLL with the KL terms (default True)
+    """
+    def __init__(self, likelihood, model, num_data, beta=1., combine_terms=True):
+        super().__init__(likelihood, model)
+        self.combine_terms = combine_terms
+        self.num_data = num_data
+        self.beta = beta
+
+    @abstractmethod
+    def _log_likelihood_term(self, approximate_dist_f, target, **kwargs):
+        raise NotImplementedError
+
+    def _kl_divergence(self):
```

Comment:
The predictive loss implemented like this will work fine for `CholeskyVariationalDistribution` and `MeanFieldVariationalDistribution`, but if we wanted to implement the Delta distribution we'll need the regularizer to be `\beta log p(u)`.

Maybe we should stick with mean field as the recommended one and punt on delta? I don't know how important the delta distribution version is -- it seemed like mean field usually did about as well on RMSE but much beter on NLL.

Reply:
Here's another thought: we could decompose this even further. We could add a `_variational_cross_entropy` term which computes `-\beta E_{q(u)} [ log p(u) ]` and a `_variational` term which computes `-\beta E_{q(u)} [ log q(u) ]`.

With the DeltaVariationalDistribution, this would then collapse to just `\beta log p(u)`. Thoughts?

Reply:
Actually a better idea:

We create a `gpytorch.distributions.Delta` distribution - similar to [Pyro's](http://docs.pyro.ai/en/stable/distributions.html#delta). We can register a kl_divergence between delta dists and MVNs.

Reply:
@gpleiss Yeah I like that idea

Reply:
Done!